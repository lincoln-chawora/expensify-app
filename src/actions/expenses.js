import db from "../firebase/firebase";

// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        // Create defaults, this is similar to adding them in arguments.
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;

        // Create expense object to store data in var.
        const expense = { description, note, amount, createdAt };

        // Add expense into firebase (push will generate id in fb), then dispatch that
        // expense into the redux store, using the new id from firebase.
        return db.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        }).catch((e) => {
            console.log('Expense creation failed', e);
        })
    }
}
// Actions generators

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return db.ref(`users/${uid}/expenses/${id}`).remove().then( () => {
            dispatch(removeExpense({id}))
        }).catch((e) => {
            console.log('Unable to remove expenses', e);
        });
    }
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return db.ref(`users/${uid}/expenses/${id}`).update(updates).then( () => {
            dispatch(editExpense(id, updates))
        }).catch((e) => {
            console.log('Unable to edit expense', e);
        });
    }
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return db.ref(`users/${uid}/expenses`).once('value').then( (snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })

            dispatch(setExpenses(expenses))
        }).catch((e) => {
            console.log('Unable to return expenses', e);
        });
    }
};
