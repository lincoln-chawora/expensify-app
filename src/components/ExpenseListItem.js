import React from "react";
import {Link} from "react-router-dom";
import moment from "moment";

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
    <div>
        <h3><Link to={"/edit/" + id}>{description}</Link></h3>
        <p>{amount} - {moment(createdAt).format('Do MMMM YYYY')}</p>
    </div>
)

export default ExpenseListItem;