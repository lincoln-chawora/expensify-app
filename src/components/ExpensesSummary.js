import React from "react";
import { connect } from 'react-redux'
import selectExpensesTotal from "../selectors/expenses-total";
import selectExpenses from "../selectors/expenses";
import numeral from "numeral";
import locales from "numeral/locales/en-gb"
import * as constants from "constants";

numeral.locale('en-gb');

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
    return (
        <div>
            <h1>Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}</h1>
        </div>
    )
}

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters)
    return {
        expensesTotal: selectExpensesTotal(visibleExpenses),
        expenseCount: visibleExpenses.length
    }
};

export default connect(mapStateToProps)(ExpensesSummary);
