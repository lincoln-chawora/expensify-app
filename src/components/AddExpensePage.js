import React from "react";
import ExpenseForm from "./ExpenseForm";
import {connect} from "react-redux";
import {startAddExpense} from "../actions/expenses";
import Heading from "./Heading";

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // Submit expense data.
        this.props.startAddExpense(expense);
        // Redirect to home page.
        this.props.history.push('/');
    };

    render() {
       return (
            <div>
                <Heading title="Add Expense" />
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
  return {
      startAddExpense: (expense) => dispatch(startAddExpense(expense))
  }
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage);