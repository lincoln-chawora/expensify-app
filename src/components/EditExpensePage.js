import React from "react";
import {connect} from "react-redux";
import ExpenseForm from "./ExpenseForm";
import {startAddExpense, startRemoveExpense, startEditExpense} from "../actions/expenses";
import Heading from "./Heading";

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        if (this.props.expense) {
            const expenseId = this.props.expense.id;
            this.props.startEditExpense(expenseId, expense)
        } else {
            this.props.startAddExpense(expense);
        }

        // Redirect to dashboard on from submit. (prop.history comes from react router)
        this.props.history.push('/');
    }
    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <Heading title="Edit Expense" />
                <div className="container">

                    <ExpenseForm
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                    />

                    <button className="btn secondary-btn" onClick={this.onRemove}>Delete Expense</button>
                </div>
            </div>
        )
    }
}
const mapStateToProps = (state, props) => {
    return {
        // State has all expenses, we find the expense that matches the url id from props.
        // Props in this context comes from react-router-dom which has access to url / page info.
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        })
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
        startAddExpense: (expense) => dispatch(startAddExpense(expense)),
        startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);