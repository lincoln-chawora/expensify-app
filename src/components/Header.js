import React from "react";
import {NavLink} from "react-router-dom";
import {startLogout} from "../actions/auth";
import {connect} from "react-redux";

export const Header = ({ startLogout }) => (
    <header>
        <h1>Expensify</h1>
        <nav>
            <NavLink to="/dashboard" activeClassName="active">Dashboard</NavLink>
            <NavLink to="/create" activeClassName="active">Create expense</NavLink>
            <button onClick={startLogout}>Logout</button>
        </nav>
    </header>
)

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
})
export default connect(undefined, mapDispatchToProps)(Header);