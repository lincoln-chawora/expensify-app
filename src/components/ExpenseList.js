import React from "react";
import { connect } from 'react-redux'
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses"

export const ExpenseList = (props) => (
    <div className="expenses-list">
        <div className="expenses-list--header">
            <div>Expense</div>
            <div>Amount</div>
        </div>
        <div className="expenses-list--content">
            {
                props.expenses.length === 0 ? (
                    <p className="empty">No expenses</p>
                ) : (
                    props.expenses.map((expense, index) => (
                        <ExpenseListItem key={expense.id} {...expense}/>
                    ))
                )
            }
        </div>
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

export default connect(mapStateToProps)(ExpenseList);
