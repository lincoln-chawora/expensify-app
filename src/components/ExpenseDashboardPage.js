import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesSummary />
        <div className="container">
            <ExpenseListFilters />
            <ExpenseList />
        </div>
    </div>
)

export default ExpenseDashboardPage