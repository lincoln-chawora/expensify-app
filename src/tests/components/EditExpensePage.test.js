import React from "react";
import {shallow} from "enzyme";
import expenses from "../fixtures/expenses";
import {EditExpensePage} from "../../components/EditExpensePage";
let startAddExpense, startRemoveExpense, startEditExpense, history, wrapper;

beforeEach(() => {
    startEditExpense = jest.fn();
    startAddExpense = jest.fn();
    startRemoveExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<EditExpensePage
        startEditExpense={startEditExpense}
        startAddExpense={startAddExpense}
        startRemoveExpense={startRemoveExpense}
        expense={expenses[0]}
        history={history}
    />)
})

test('Should render edit expense page directly', () => {
    expect(wrapper).toMatchSnapshot();
})

test('Should handle edit submit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startEditExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
})

test('Should handle remove expense', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith({
        id: expenses[0].id
    });
})