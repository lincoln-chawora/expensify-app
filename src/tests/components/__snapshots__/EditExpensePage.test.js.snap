// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render edit expense page directly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    className="remove-expense"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
